service_container/3.3-di-changes.rst

```diff
@@ -0,0 +1,740 @@
+            # and have a tag that allows actions to type-hint services
+            AppBundle\Controller\:
+                resource: '../../src/AppBundle/Controller'
+                public: true
+                tags: ['controller.service_arguments']
+
+    .. code-block:: xml
+
+        <!-- app/config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <!-- ... -->
+
+                <prototype namespace="AppBundle\Controller" resource="../../src/AppBundle/Controller" public="true">
+                    <tag name="controller.service_arguments" />
+                </prototype>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // app/config/services.php
+
+        // loading entire directories is not possible with PHP configuration
+        // you need to define your services one-by-one
+        use AppBundle\Controller\DefaultController;
+
+        $container->autowire(DefaultController::class)
+            ->setAutoconfigured(true)
+            ->addTag('controller.service_arguments')
+            ->setPublic(false);
+
+But, you might not even notice this. First, your controllers *can* still extend
+the same base ``Controller`` class or a new :ref:`AbstractController <controller-abstract-versus-controller>`.
+This means you have access to all of the same shortcuts as before. Additionally,
+the ``@Route`` annotation and ``_controller`` syntax (e.g.``AppBundle:Default:homepage``)
+used in routing will automatically use your controller as a service (as long as its
+service id matches its class name, which it *does* in this case).
+
+In other words, everything works the same. You can even add the above configuration
+to your existing project without any issues: your controllers will behave the same
+as before. But now that your controllers are services, you can use dependency injection
+and autowiring like any other service.
+
+To make life even easier, it's now possible to autowire arguments to your controller
+action methods, just like you can with the constructor of services. For example::
+
+    use Psr\Log\LoggerInterface;
+
+    class InvoiceController extends Controller
+    {
+        public function listAction(LoggerInterface $logger)
+        {
+            $logger->info('A new way to access services!');
+        }
+    }
+
+This is special to controllers *only*, and your controller service must be tagged
+with ``controller.service_arguments`` to make it happen. This new method is used
+throughout the documentation. You can use it, use normal constructor dependency injection,
+or continue to fetch *public* services via ``$this->container->get()``.
```

Comment:
should this be rephrased with something like "The new best practice is now to use it or use normal constructor dependency injection, instead of fetching *public* services via ..., which still works"?